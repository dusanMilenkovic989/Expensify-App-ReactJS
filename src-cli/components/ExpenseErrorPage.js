import React from 'react'
import { Link } from 'react-router-dom'

const ExpenseErrorPage = () => (
    <div>
        <p>404 Page</p>
        <Link to="/dashboard">Go home</Link>
    </div>
)

export { ExpenseErrorPage as default }