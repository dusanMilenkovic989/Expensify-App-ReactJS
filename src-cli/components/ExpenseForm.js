import React from 'react'
import { SingleDatePicker } from 'react-dates'
import moment from 'moment'

class ExpenseForm extends React.Component {
    state = {
        description: this.props.expense ? this.props.expense.description : '',
        amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
        note: this.props.expense ? this.props.expense.note : '',
        createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
        calendarFocused: false,
        error: '' 
    }
    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) this.setState(() => ({ amount }))
    }
    onDateChange = (createdAt) => {
        if (createdAt) this.setState(() => ({ createdAt }))
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }
    onSubmit = (e) => {
        e.preventDefault()

        if (!this.state.description || !this.state.amount) return this.setState(() => ({ error: 'Description and amount must be provided!' }))

        this.setState(() => ({ error: '' }))
        this.props.onFormSubmit({
            description: this.state.description,
            amount: parseFloat(this.state.amount, 10) * 100,
            note: this.state.note,
            createdAt: this.state.createdAt.valueOf()
        })
    }
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="Description"
                    autoFocus
                    className="text-input"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="text"
                    placeholder="Amount"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <div>
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                </div>
                <textarea
                    placeholder="Add expense note here"
                    className="textarea"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                />
                <div>
                    <button className="button">{this.props.expense ? 'Update Expense' : 'Add Expense'}</button>
                </div>
            </form>
        )
    }
}

export { ExpenseForm as default }