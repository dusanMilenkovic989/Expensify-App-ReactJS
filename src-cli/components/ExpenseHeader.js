import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import visibleExpenses from '../redux/selectors/expenses'
import summary from '../redux/selectors/summary'
import numeral from '../utils/numeral'

export const ExpenseHeader = ({ expenses }) => {
    const plural = expenses.length !== 1 ? 's' : ''
    const message = (
        <h1 className="page-header__title">
            Viewing <span>{expenses.length}</span> expense{plural}, totalling <span>{numeral(summary(expenses) / 100).format('0,0.00 $')}</span>
        </h1>
    )

    return (
        <div className="page-header">
            <div className="content-container">  
                {expenses.length > 0 && message}
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    )
}

const mapStateToProps = (state) => ({ expenses: visibleExpenses(state.expenses, state.filters) })

export default connect(mapStateToProps)(ExpenseHeader)