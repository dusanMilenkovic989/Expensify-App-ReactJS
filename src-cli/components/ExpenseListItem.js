import React from 'react'
import { Link } from 'react-router-dom'
import moment from 'moment'
import numeral from '../utils/numeral'

const ExpenseListItem = ({ expenseIndex, id, description, amount, createdAt }) => (
    <Link className="list-item" to={`/edit/${id}`}>
        <div>
            <h3 className="list-item__title">{expenseIndex}. {description}</h3>
            <span className="list-item__subtitle">{moment(createdAt).format('Do MMMM, YYYY')}</span>
        </div>
        <h3 className="list-item__data">{numeral(amount / 100).format('0,0.00 $')} </h3>
    </Link>
)

export { ExpenseListItem as default }