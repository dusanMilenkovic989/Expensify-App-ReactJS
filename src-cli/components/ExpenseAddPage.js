import React from 'react'
import { useNavigate } from 'react-router-dom'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { addExpense } from '../redux/actions/expenses'

const withNavigateExpenseAddPage = (props) => {
    const navigate = useNavigate()

    return <ExpenseAddPage {...props} navigate={navigate} />
}

export const ExpenseAddPage = ({ addExpense, navigate }) => (
    <div>
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Add Expense</h1>
            </div>
        </div>
        <div className="content-container">
            <ExpenseForm 
                onFormSubmit={(expense) => { 
                    addExpense(expense)
                    navigate('/')
                }}
            />
        </div>
    </div>
)

const mapDispatchToProps = (dispatch) => ({ addExpense: (expense) => { dispatch(addExpense(expense)) } })

export default connect(undefined, mapDispatchToProps)(withNavigateExpenseAddPage)