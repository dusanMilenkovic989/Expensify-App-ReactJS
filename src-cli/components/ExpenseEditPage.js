import React from 'react'
import { useNavigate, useParams } from 'react-router-dom'
import { connect } from 'react-redux'
import { editExpense, removeExpense } from '../redux/actions/expenses'
import ExpenseForm from './ExpenseForm'

const withParAndNavExpenseEditPage = (props) => {
    const params = useParams()
    const navigate = useNavigate()

    return <ExpenseEditPage {...props} params={params} navigate={navigate} />
}

export const ExpenseEditPage = ({ expenses, editExpense, removeExpense, params, navigate }) => (
    <div>
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Edit Expense</h1>
            </div>
        </div>
        <div className="content-container">
            <ExpenseForm 
                onFormSubmit={(expense) => { 
                    editExpense(params.id, expense)
                    navigate('/')
                }}
                expense={expenses.find((expense) => expense.id === params.id)}
            />
            <button
                className="button button--secondary"
                onClick={() => { 
                    removeExpense({ id: params.id })
                    navigate('/')
                }}
            >
                Remove expense
            </button>
        </div>
    </div>
)

const mapStateToProps = (state) => ({ expenses: state.expenses })

const mapDispatchToProps = (dispatch) => ({ 
    editExpense: (expenseId, expense) => { dispatch(editExpense(expenseId, expense)) },
    removeExpense: (data) => { dispatch(removeExpense(data)) }
})

export default connect(mapStateToProps, mapDispatchToProps)(withParAndNavExpenseEditPage)