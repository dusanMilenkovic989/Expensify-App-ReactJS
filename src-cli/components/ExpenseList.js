import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import VisibleExpenses from '../redux/selectors/expenses'

export const ExpenseList = ({ expenses }) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                expenses.length > 0 ? (
                    expenses.map((expense) => ( 
                        <ExpenseListItem key={expense.id} expenseIndex={expenses.indexOf(expense) + 1} {...expense} />
                    ))
                ) : (
                    <div className="list-item list-item--message">
                        <span>No expenses yet</span>
                    </div>
                )
            }
        </div>
    </div>
)

const mapStateToProps = (state) => ({ expenses: VisibleExpenses(state.expenses, state.filters) })

export default connect(mapStateToProps)(ExpenseList)