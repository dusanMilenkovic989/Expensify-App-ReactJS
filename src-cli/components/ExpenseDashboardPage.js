import React from 'react'
import ExpenseHeader from './ExpenseHeader'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseList from './ExpenseList'

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseHeader />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export { ExpenseDashboardPage as default }