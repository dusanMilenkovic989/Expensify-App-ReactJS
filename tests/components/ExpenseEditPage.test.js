import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseEditPage } from '../../src-cli/components/ExpenseEditPage'
import { expenses } from '../fixtures/expenses-and-filters'

let editExpense, removeExpense, params, navigate, wrapper

beforeEach(() => {
    editExpense = jest.fn()
    removeExpense = jest.fn()
    params = { id: expenses[2].id }
    navigate = jest.fn()
    wrapper = shallow(
        <ExpenseEditPage 
            expenses={expenses}
            editExpense={editExpense}
            removeExpense={removeExpense}
            params={params}
            navigate={navigate}
        />
    )
})

test('Should render ExpenseEditPage correctly', () => {
    expect(wrapper).toMatchSnapshot()
})

test('Should dispatch editExpense and navigate', () => {
    wrapper.find('ExpenseForm').prop('onFormSubmit')(expenses[2])

    expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2])
    expect(navigate).toHaveBeenLastCalledWith('/')
})

test('Should dispatch removeExpense and navigate', () => {
    wrapper.find('button').simulate('click')

    expect(removeExpense).toHaveBeenLastCalledWith({ id: expenses[2].id })
    expect(navigate).toHaveBeenLastCalledWith('/')
})