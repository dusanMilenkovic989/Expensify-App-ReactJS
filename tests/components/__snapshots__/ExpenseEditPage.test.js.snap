// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseEditPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 750,
          "createdAt": 345600000,
          "description": "Burger",
          "id": "3",
          "note": "",
        }
      }
      onFormSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove expense
    </button>
  </div>
</div>
`;
